src/index.ts

```diff
@@ -1,65 +1,120 @@
+				return new Response('Durable Object expected Upgrade: websocket', {
+					status: 426,
+				});
+			}
+
+			// This example will refer to the same Durable Object,
+			// since the name "foo" is hardcoded.
+			let id = env.WEBSOCKET_HIBERNATION_SERVER.idFromName('foo');
+			let stub = env.WEBSOCKET_HIBERNATION_SERVER.get(id);
+
+			return stub.fetch(request);
+		}
+
+		return new Response(null, {
+			status: 400,
+			statusText: 'Bad Request',
+			headers: {
+				'Content-Type': 'text/plain',
+			},
+		});
+	},
+};
+
+// Durable Object
 export class WebSocketHibernationServer extends DurableObject {
-	/**
-	 * The constructor is invoked once upon creation of the Durable Object, i.e. the first call to
-	 * 	`DurableObjectStub::get` for a given identifier (no-op constructors can be omitted)
-	 *
-	 * @param ctx - The interface for interacting with Durable Object state
-	 * @param env - The interface to reference bindings declared in wrangler.toml
-	 */
+	// Keep track of all WebSocket connections
+	sessions: Map<WebSocket, any>;
+
 	constructor(ctx: DurableObjectState, env: Env) {
 		super(ctx, env);
+		this.sessions = new Map();
+
+		// DO saves the connection information when it goes to hibernation
+		// Get all WebSocket connections from the DO
+		this.ctx.getWebSockets().forEach((ws) => {
+			// Checks if this is a new DO
+			if (ws.deserializeAttachment()) {
+				// If a DO is waking up, deserialize the attachment.
+				// This helps to restore the state of the connection
+				const { ...session } = ws.deserializeAttachment();
+				this.sessions.set(ws, { ...session });
+			}
+		});
+	}
+
+	async fetch(request: Request): Promise<Response> {
+		// Creates two ends of a WebSocket connection.
+		const webSocketPair = new WebSocketPair();
+		const [client, server] = Object.values(webSocketPair);
+
+		// Calling `acceptWebSocket()` informs the runtime that this WebSocket is to begin terminating
+		// request within the Durable Object. It has the effect of "accepting" the connection,
+		// and allowing the WebSocket to send and receive messages.
+		// Unlike `ws.accept()`, `state.acceptWebSocket(ws)` informs the Workers Runtime that the WebSocket
+		// is "hibernatable", so the runtime does not need to pin this Durable Object to memory while
+		// the connection is open. During periods of inactivity, the Durable Object can be evicted
+		// from memory, but the WebSocket connection will remain open. If at some later point the
+		// WebSocket receives a message, the runtime will recreate the Durable Object
+		// (run the `constructor`) and deliver the message to the appropriate handler.
+		this.ctx.acceptWebSocket(server);
+
+		// Generate a random UUID for the session.
+		const id = crypto.randomUUID();
+
+		// Attach the session ID to the WebSocket connection and serialize it.
+		// This is necessary to restore the state of the connection when the Durable Object wakes up.
+		server.serializeAttachment({
+			...server.deserializeAttachment(),
```

Comment:
Maybe I'm missing something, but I cannot think of a case where server would have an attachment at this stage. So I don't believe that this line is needed.

Reply:
I was trying to address this comment:  https://github.com/cloudflare/cloudflare-docs/pull/17670/files#r1808955786

if this isn't the correct approach, where should the ID be generated and attached to the connection?

Reply:
Going through the chat example, it looks like we are doing the same there, but in the `handleSession` method.
https://github.com/cloudflare/workers-chat-demo/blob/97bacfa773c673ef35f974dc076ba65c42fede71/src/chat.mjs#L303

Reply:
@harshil1712 I actually have the context for doing that [here](https://github.com/cloudflare/workers-chat-demo/pull/32#discussion_r1198304406).

The concern is basically just if someone were to add more stuff to the attachment and call `serializeAttachment()` multiple times, it could break.

I don't think it's too important here though, we can just do `server.serializeAttachment({id});`